ext/drstatecmp/drstatecmp.dox

```diff
@@ -0,0 +1,146 @@
+\page page_drstatecmp Machine State Comparison Library
+
+The \p drstatecmp DynamoRIO Machine State Comparison Extension provides mechanisms
+to enable systematic and exhaustive machine state comparisons across instrumentation
+sequences. These comparisons will detect instrumentation-induced corruptions of the
+application's machine state that would typically lead to obscure bugs.
+
+ - \ref sec_drstatecmp_init
+ - \ref sec_drstatecmp_checks
+ - \ref sec_drstatecmp_instr_state_sharing
+ - \ref sec_drstatecmp_machine_state_saved
+ - \ref sec_drstatecmp_vs_drbbdup
+
+\section sec_drstatecmp_init Setup
+
+To use \p drstatecmp with your client simply include this line in your client's
+\p CMakeLists.txt file:
+
+\code use_DynamoRIO_extension(clientname drstatecmp) \endcode
+
+That will automatically set up the include path and library dependence.
+
+The initialization routine \p drstatecmp_init() must be called prior to any of
+the other routines.
+
+\section sec_drstatecmp_checks Checks
+
+There are two types of checks depending on whether a basic block has any side-effects.
+
+\subsection sec_drstatecmp_checks_side_effects_free Side-effect-free Basic Blocks
+
+Side-effect-free basic blocks are duplicated and executed twice, once with the full
+instrumentation sequence and once with no instrumentation except for a state comparison at
+the end. Essentially, we check whether the machine state at the end of the basic block is
+the same with and without instrumentation code. This check should catch any state
+clobbering by instrumentation.
+
+The side-effect-free basic blocks are duplicated at drmgr's app2app stage. The copy of
+each such basic block is appended immediately after the original basic block. Labels are
+inserted in the beginning of the original basic block (ORIG_BB) and its copy (COPY_BB) to
+allow identification in later stages. The terminating instruction of the basic block (if
+any) is only kept in the copy and is preceded by a TERM label. A jump to the TERM label is
+inserted at the end the original basic block to skip the duplicated bb.
+
+The insert instrumentation stage inserts the actual checks. First, the original basic
+block is instrumented. In the beginning of the original basic block, the machine state is
+saved. This state will be used for state restoration before the re-execution of the basic
+block. At the end of the original block, code is inserted to save the state for later
+comparison and restore the machine state to the state before executing this bb. This stage
+also removes the previously inserted jump instr to the TERM label and let it fall through
+to the instructions of the copy of the basic block. Then, the copy of the basic block is
+instrumented. At the end of the copy, a clean call is inserted that compares the current
+machine state with the machine state that was saved at the end of the original basic
+block. Finally, all previously-inserted (from other passes) instrumentation instructions
```

Comment:
You mean from other app2app actions like rep string expansion?  How do you know how to undo such expansion?  It seems like you would instead need to have a high priority app2app callback to copy before anyone changed the code?